haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+qualifier. Attempts to place the new import next to existing imports of the same
+module if they exist."
+  (interactive "MModule Name: \nMQualifier: ")
+  (let* ((all-infos (haskell-auto-import-module-import-info module-name))
+         (qualified
+          (member-if '(lambda (info)
+                        (car (cdr (cdr info))))
+                     all-infos))
+         (same-qualifier
+          (member-if '(lambda (info)
+                        (string= (car (cdr (cdr info))) qualifier))
+                     all-infos)))
+    (if same-qualifier
+        (message "That module is already imported with the same qualifier. Doing nothing.")
+      (let ((old-point (point)))
+        ;; If the module is already imported in some capacity, put the new
+        ;; import above the existing ones
+        (if all-infos
+            (goto-char (+ 1 (car (car all-infos))))
+          (haskell-navigate-imports))
+
+        (insert
+         (haskell-auto-import-print-module-import
+          module-name
+          (list t qualifier nil)))
+        (newline)
+
+        (goto-char old-point)))))
+
+(defun haskell-auto-import-qualified-and-unqualified (module-name qualifier sub-imports)
+  "Add a new import, both qualified and unqualified. For example, given the
+module `Data.Map', qualifier `Map' and sub-imports `Map', this will add the lines:
+
+import Data.Map (Map)
+import qualified Data.Map as Map"
+
+  (interactive
+   "MModule Name: \nMQualifier: \nMSub-Imports (comma separated, leave blank to import all): ")
+  (haskell-auto-import-qualified module-name qualifier)
+  (haskell-auto-import-unqualified module-name sub-imports))
+
+(defun haskell-auto-import-parse-import-parens (contents)
+  "Given the contents of some import parentheses, return a list of pairs which
+each consist of '(imported-item sub-imports), where sub-imports is the list of
+items imported from the main item (e.g. in `import ModuleName(DataType(Con1,
+Con2))')
+
+Does NOT currently work properly with datatype operators with sub-imports (e.g.
+`import Foo.Bar ((**)(A, B))') or with sub-imports which are operators (e.g.
+`import Foo.Bar (A((&&), (||)))') "
+  (let* ((exploded (split-string contents "," nil "\\s-"))
+         (parts
+          (reduce
+           '(lambda (acc str)
+              (let ((cur (car acc))
+                    (rest (cdr acc))
+                    (strlist (string-to-list str)))
+                (if cur
+                    (if (member ?\) strlist)
+                        (cons () (cons (cons str cur) rest))
+                      (cons (cons str cur) rest))
+                  (if (and
+                       (member ?\( strlist)
+                       ;; If the close paren is present too, we have a whole
+                       ;; token
+                       (not (member ?\) strlist)))
+                      (cons (list str) rest)
+                    (cons () (cons (list str) rest))))))
```

Comment:
It's better to use `nil` instead of empty list literals `()` to represent falsey values. They mean the same thing, but `nil` is easier to spot since it's visually distinct from the rest of the parens in the code structure.